resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/Client.scala

```diff
@@ -0,0 +1,215 @@
+import scala.collection.mutable
+import scala.util.control.NonFatal
+
+import io.fabric8.kubernetes.api.model._
+import io.fabric8.kubernetes.client.KubernetesClient
+
+import org.apache.spark.SparkConf
+import org.apache.spark.deploy.k8s.SparkKubernetesClientFactory
+import org.apache.spark.deploy.k8s.config._
+import org.apache.spark.deploy.k8s.constants._
+import org.apache.spark.deploy.k8s.submit.steps.{DriverConfigurationStep, KubernetesDriverSpec}
+import org.apache.spark.internal.Logging
+import org.apache.spark.util.Utils
+
+private[spark] case class ClientArguments(
+     mainAppResource: MainAppResource,
+     mainClass: String,
+     driverArgs: Array[String])
+
+private[spark] object ClientArguments {
+  def fromCommandLineArgs(args: Array[String]): ClientArguments = {
+    var mainAppResource: Option[MainAppResource] = None
+    var mainClass: Option[String] = None
+    val driverArgs = mutable.Buffer.empty[String]
+
+    args.sliding(2, 2).toList.collect {
+      case Array("--primary-java-resource", primaryJavaResource: String) =>
+        mainAppResource = Some(JavaMainAppResource(primaryJavaResource))
+      case Array("--main-class", clazz: String) =>
+        mainClass = Some(clazz)
+      case Array("--arg", arg: String) =>
+        driverArgs += arg
+      case other =>
+        val invalid = other.mkString(" ")
+        throw new RuntimeException(s"Unknown arguments: $invalid")
+    }
+
+    require(mainAppResource.isDefined,
+      "Main app resource must be defined by either --primary-py-file or --primary-java-resource.")
+    require(mainClass.isDefined, "Main class must be specified via --main-class")
+
+    ClientArguments(
+      mainAppResource.get,
+      mainClass.get,
+      driverArgs.toArray)
+  }
+}
+
+private[spark] class Client(
+    submissionSteps: Seq[DriverConfigurationStep],
+    submissionSparkConf: SparkConf,
+    kubernetesClient: KubernetesClient,
+    waitForAppCompletion: Boolean,
+    appName: String,
+    loggingPodStatusWatcher: LoggingPodStatusWatcher) extends Logging {
+
+  private val driverJavaOptions = submissionSparkConf.get(
+    org.apache.spark.internal.config.DRIVER_JAVA_OPTIONS)
+
+   /**
+    * Run command that initializes a DriverSpec that will be updated after each
+    * DriverConfigurationStep in the sequence that is passed in. The final KubernetesDriverSpec
+    * will be used to build the Driver Container, Driver Pod, and Kubernetes Resources
+    */
+  def run(): Unit = {
+    var currentDriverSpec = KubernetesDriverSpec.initialSpec(submissionSparkConf)
+    // submissionSteps contain steps necessary to take, to resolve varying
+    // client arguments that are passed in, created by orchestrator
+    for (nextStep <- submissionSteps) {
+      currentDriverSpec = nextStep.configureDriver(currentDriverSpec)
+    }
+
+    val resolvedDriverJavaOpts = currentDriverSpec
+      .driverSparkConf
+      // We don't need this anymore since we just set the JVM options on the environment
```

Comment:
Where is the code that this comment is referring to?

Reply:
Good catch. This comment and the line below it should be removed. Cc @foxish @mccheah to confirm.

Reply:
Actually it seems it was added in #365. I think we should keep the comment but rephrase it to make it clearer.

Reply:
Comment still applies because the lines below take the driver JVM options and append them to the JVM options of the SparkConf. For example, we don't want to have SPARK_DRIVER_JAVA_OPTS have a value of `-Dspark.driver.extraJavaOptions=-XX:HeapDumpOnOutOfMemoryError` - we just want the `-XX:HeapDumpOn...` to be set directly on the driver process.

Reply:
The comment should be rephrased to clarify the intent of removing `extraJavaOptions`, along the lines of what Matt said.

Reply:
Done.